### PR TITLE
[Google Blockly] Copy/Cut/Paste should ignore next block

### PR DESCRIPTION
**Pull request by mikeharv**
## Context

This PR aims to correct the behavior of context menu items and shortcuts relating to copying and cutting blocks.

### Click below to read some lengthy history
<details>
<summary> Expand for details</summary>

As part of the work to maintain our codebase related to Google Blockly, we plan to a Blockly plugin for context menu items and keyboard shortcuts to allow users to copy and paste a block between tabs: 
- https://npm.io/package/@blockly/plugin-cross-tab-copy-paste

When we first recognized the need for this functionality, the plugin was only available for v7 Blockly. As we were still on v6, we chose to manually implement (re-write) the same behavior found in the plugin in a way that was compatible:
- https://github.com/code-dot-org/code-dot-org/pull/45862

Next, we began prepping the work to revert the above PR in favor of the using the plugin. Red PRs are great!
- https://github.com/code-dot-org/code-dot-org/pull/48997
Unfortunately, we've learned that this version of the plugin expects v9 of Blockly so we are currently blocked from merging. 

While working on this, we noticed an inconsistency between the plugin's behavior and custom rewrite: We were including blocks connected to a [next connection](https://developers.google.com/blockly/guides/create-custom-blocks/define-blocks#next_connection) with each copy/cut/paste command. As it turns out, this choice we made this decision based on a blindspot we had about how blocks can be deleted. This was fixed:
- https://github.com/code-dot-org/code-dot-org/pull/49073

Once merged, we realized that the behavior for 'cut' no longer matched with how delete was now going to work. Futhermore, 'copy' was also still behaving in a way that was inconsistent with core Google Blockly as well as CDO Blockly labs. This leaves us with two options:
1. Race to bump to v9 (high risk, unknown difficulty) so that we can start using the plugin ASAP
2. Correct our copy/cut logic to be more consistent now and worry about the plugin later.

This PR follows option #2.

</details>

|Task|Before|After|
|-|-|-|
|**Context menu** - Copy a single block<br> *No change*|1b| 1a|
|**Context menu** - Copy a block with with a next block<br>*Next block should not be pasted.*|2b | 2a|
|**Shortcuts** - Copy a single block<br> *No change*| 3b|3a |
|**Shortcuts** - Copy a block with with a next block<br>*Next block should not be pasted.*|4b | 4a|
|**Shortcuts** - Cut a single block<br> *No change*|5b | 5a|
|**Shortcuts** - Cut a block with with a next block<br>*Next block should not be pasted.*| 6b| 6a|

I also checked that in each of the above cases we were still correctly storing XML for any child blocks.
While each above scenario will not be illustrated, here is an example:

## Testing story

I think it would be fairly straight-forward to implement a unit test to check the functionality of `getSingleBlockXml`. I haven't done that yet because I'd like to check this direction first. I'm also wondering if it's worth adding a test for this, knowing that we'd plan to delete it once the plugin is finally added.

## Follow-up work

After bumping to v9 of Google Blockly, we can revert this branch as well as merge this one: https://github.com/code-dot-org/code-dot-org/pull/48997

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/addons/contextMenu.js [hunk 1]
```diff
@@ -18,9 +19,7 @@ const registerBlockCopyToStorage = function() {
       }
     },
     callback: function(scope) {
-      const copyData = Blockly.Xml.domToPrettyText(
-        Blockly.Xml.blockToDom(scope.block)
-      );
+      let copyData = getSingleBlockXml(scope.block);
```
Comment (bencodeorg) [inline apps/src/blockly/addons/contextMenu.js]: Nit: const

File: apps/src/blockly/addons/cdoUtils.js [hunk 2]
```diff
@@ -89,3 +89,29 @@ export function getField(type) {
...
+ * @param {block}
+ * @returns {string}
+ */
+export function getSingleBlockXml(block) {
+  let blockStackXml = Blockly.Xml.domToPrettyText(
+    Blockly.Xml.blockToDom(block)
+  );
+  // Remove whitespace from indentation.
+  blockStackXml = blockStackXml.replace(/\s+/g, ' ');
+  let blockXml = blockStackXml;
```
Comment (bencodeorg) [inline apps/src/blockly/addons/cdoUtils.js]: could this just be `let blockXml = blockStackXml.replace(/\s+/g, ' ');`?

**Review by bencodeorg — APPROVED** [review]

File: apps/src/blockly/googleBlocklyWrapper.js [hunk 3]
```diff
@@ -272,7 +272,8 @@ function initializeBlocklyWrapper(blocklyInstance) {
     this.removeUnusedBlockFrame();
   };
 
-  // overridden dispose function definition at https://github.com/google/blockly/blob/develop/core/block_svg.ts#L863
+  // The original Google Blockly dispose() is defined at:
+  // https://github.com/google/blockly/blob/1f862cb878f7eec36b71c638b85d5199bff01fcb/core/block_svg.ts#L863
```
Comment (fisher-alice) [inline apps/src/blockly/googleBlocklyWrapper.js:276]: Thanks for replacing link with permalink!

File: apps/src/blockly/addons/cdoUtils.js [hunk 4]
```diff
@@ -89,3 +89,29 @@ export function getField(type) {
...
+
+/**
+ * Returns xml string for a block, after removing any next block
+ * connected to it. Used by context menu options and shortcuts.
+ * This should not be used elsewhere as it is scheduled for deletion
+ * once we implement @blockly/plugin-cross-tab-copy-paste
+ * @param {block}
+ * @returns {string}
+ */
+export function getSingleBlockXml(block) {
```
Comment (fisher-alice) [inline apps/src/blockly/addons/cdoUtils.js]: I know this is a time-sensitive PR, but just curious about why we need to add this helper function to achieve this effect when Google Blockly's copy and cut behavior already behaves like this. Is it possible to use their `copy` which doesn't include [next blocks](https://github.com/google/blockly/blob/92efdb91d12281a412aac4e87701471b80406393/core/block_svg.ts#L944) as we use their `dispose`?

**Review by fisher-alice — APPROVED** [review]
Nice job on this! I was impressed by the table of before/after  screencast GIFs! Thanks

File: apps/src/blockly/addons/cdoUtils.js (hunk 4 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js]: ALICE! You were so right. Check out how much cleaner this is now that I'm using the built-in `toCopyData` and `pasteBlock_` (https://github.com/google/blockly/blob/34634ea57cd43dac9028f1a28e4beb761e60eef6/core/workspace_svg.ts#L1400).

File: apps/src/blockly/addons/contextMenu.js [hunk 5]
```diff
@@ -18,9 +18,7 @@ const registerBlockCopyToStorage = function() {
       }
     },
     callback: function(scope) {
-      const copyData = Blockly.Xml.domToPrettyText(
-        Blockly.Xml.blockToDom(scope.block)
-      );
+      let copyData = JSON.stringify(Blockly.selected.toCopyData().saveInfo);
```
Comment (fisher-alice) [inline apps/src/blockly/addons/contextMenu.js]: nit: use const as you do below.

File: apps/src/blockly/addons/shortcut.js [hunk 6]
```diff
@@ -22,9 +22,7 @@ const blockCopyToStorageShortcut = function() {
       // which may beep or otherwise indicate
       // an error due to the lack of a selection.
       e.preventDefault();
-      const blockText = Blockly.Xml.domToPrettyText(
-        Blockly.Xml.blockToDom(Blockly.selected)
-      );
+      const blockText = JSON.stringify(Blockly.selected.toCopyData().saveInfo);
```
Comment (fisher-alice) [inline apps/src/blockly/addons/shortcut.js]: small nit: you use a different local variable name here vs. in contextMenu to store copied data (blockText vs copyData)

File: apps/src/blockly/addons/contextMenu.js [hunk 7]
```diff
@@ -49,10 +47,7 @@ const registerBlockPasteFromStorage = function() {
     },
     callback: function(scope) {
       const copyData = localStorage.getItem('blocklyStash');
-      Blockly.Xml.domToBlockSpace(
-        Blockly.mainBlockSpace,
-        Blockly.Xml.textToDom(`<xml>${copyData}</xml>`)
-      );
+      Blockly.mainBlockSpace.pasteBlock_(false, JSON.parse(copyData));
```
Comment (fisher-alice) [inline apps/src/blockly/addons/contextMenu.js]: Curious why you're calling on private pasteBlock_ function here vs public paste? Could use `Blockly.mainBlockSpace.paste(JSON.parse(copyData))`

**Review by fisher-alice — APPROVED** [review]
This is a great refactoring! Just left a couple small comments.

File: apps/src/blockly/addons/cdoUtils.js (hunk 4 above)
Comment (fisher-alice) [reply apps/src/blockly/addons/cdoUtils.js]: Cool!

**Pull request merged**